core/src/prediction/RegressionPredictionStrategy.cpp

```diff
@@ -97,14 +97,21 @@ PredictionValues RegressionPredictionStrategy::precompute_prediction_values(
       continue;
     }
 
-    std::vector<double>& averages = values[i];
-    averages.resize(1);
-
-    double average = 0.0;
+    double sum = 0.0;
+    double weight = 0.0;
     for (auto& sample : leaf_node) {
-      average += data->get_outcome(sample);
+      sum += data->get_weight(sample) * data->get_outcome(sample);
+      weight  += data->get_weight(sample);
     }
-    averages[OUTCOME] = average / leaf_node.size();
+
+    // if total weight is zero, treat the leaf as empty
+    if (weight == 0) {
```

Comment:
I alluded to this above, but we can't actually use `==` when checking if a floating point is equal to zero. We'll need to do something like the following: `if (std::abs(weight) < 1e-10) { ... }`.

This got me thinking, should we ensure that the weights sum to 1? This will protect against overflow in calculating `sum` and `weight`. Maybe we can think about this for a follow-up.